packages/flutter_tools/lib/src/runner/flutter_command.dart

```diff
@@ -216,23 +216,60 @@ abstract class FlutterCommand extends Command<void> {
   /// Adds options for connecting to the Dart VM observatory port.
   void usesPortOptions() {
     argParser.addOption(observatoryPortOption,
-        help: 'Listen to the given port for an observatory debugger connection.\n'
+        help: '(deprecated use host-vmservice-port instead)'
+              'Listen to the given port for an observatory debugger connection.\n'
               'Specifying port 0 (the default) will find a random free port.',
     );
+    argParser.addOption('device-vmservice-port',
+      help: 'Look for vmservice connections only from the specified port.\n'
+            'Specifying port 0 (the default) will accept the first vmservice '
+            'discovered.',
+    );
+    argParser.addOption('host-vmservice-port',
+      help: 'When a device-side vmservice port is forwarded to a host-side '
+            'port, use this value as the host port.\nSpecifying port 0 '
+            '(the default) will find a random free host port.'
+    );
     _usesPortOption = true;
   }
 
-  /// Gets the observatory port provided to in the 'observatory-port' option.
+  /// Gets the vmservice port provided to in the 'observatory-port' or
+  /// 'host-vmservice-port option.
+  ///
+  /// Only one of "host-vmservice-port" and "observatory-port" may be
+  /// specified.
+  ///
+  /// If no port is set, returns null.
+  int get hostVmservicePort {
+    if (!_usesPortOption ||
+        (argResults['observatory-port'] == null &&
```

Comment:
The bug was here: This was an `||` before, which meant this early returned as null if only one was passed